<a href="https://colab.research.google.com/github/michhottinger/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/3_Michelle_Hottinger_LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [29]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-10-02 21:20:53--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.107.22
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.107.22|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  44.3MB/s    in 4.9s    

2019-10-02 21:20:58 (40.4 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [30]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [31]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01/instacart_2017_05_01


In [32]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [33]:
##### YOUR CODE HERE #####
##contains which product
import pandas as pd
prod = pd.read_csv('products.csv')
prod.head()
#contains product ID and product name

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [34]:
orders = pd.read_csv('orders.csv')
orders.head()
##contains order_id and order number

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [35]:
train = pd.read_csv('order_products__train.csv')
train.head()
##need add to card order
#order ID
##product ID  and order ID

order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [36]:
prior = pd.read_csv('order_products__prior.csv')
prior.head()
#need add_to_cart_order
#order ID prod ID
#product id frequency
#product id to product name, see how many instances a product id is used.

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

product name + Prod ID, Prod ID+ Order ID, Order ID+ Order Number

In [37]:
aisles= pd.read_csv('aisles.csv')
aisles.head()
#not usefule
dep = pd.read_csv('departments.csv')
dep.head()
##not useful

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [53]:
order_prod = pd.concat([train, prior], axis=0, join='inner')
order_prod.shape
#order_prod.head(20)

(33819106, 4)

In [0]:
#assert (train.shape[0] + prior.shape[0] == order_prod.shape[0])
#hmm didn't work?

In [40]:
#(order_products['order_id'] == 2539329) | (order_products['order_id'] == 2398795) 

prod[(prod['product_name'] == 'Banana') | 
     (prod['product_name'] == 'Bag of Organic Bananas') | 
     (prod['product_name'] == 'Organic Strawberries') | 
     (prod['product_name'] == 'Organic Baby Spinach') | 
     (prod['product_name'] == 'Organic Hass Avocado') | 
     (prod['product_name'] == 'Organic Avocado') | 
     (prod['product_name'] == 'Large Lemon') | 
     (prod['product_name'] == 'Strawberries') | 
     (prod['product_name'] == 'Limes') | 
     (prod['product_name'] == 'Organic Whole Milk')].head(20)



product_id            product_name  aisle_id  department_id
13175       13176  Bag of Organic Bananas        24              4
16796       16797            Strawberries        24              4
21136       21137    Organic Strawberries        24              4
21902       21903    Organic Baby Spinach       123              4
24851       24852                  Banana        24              4
26208       26209                   Limes        24              4
27844       27845      Organic Whole Milk        84             16
47208       47209    Organic Hass Avocado        24              4
47625       47626             Large Lemon        24              4
47765       47766         Organic Avocado        24              4

In [54]:

condition = ((prod['product_name'] == 'Banana') | 
    (prod['product_name'] == 'Bag of Organic Bananas') | 
    (prod['product_name'] == 'Organic Strawberries') | 
    (prod['product_name'] == 'Organic Baby Spinach') | 
    (prod['product_name'] == 'Organic Hass Avocado') | 
    (prod['product_name'] == 'Organic Avocado') | 
    (prod['product_name'] == 'Large Lemon') | 
    (prod['product_name'] == 'Strawberries') | 
    (prod['product_name'] == 'Limes') | 
    (prod['product_name'] == 'Organic Whole Milk'))
columns = ['product_id', 'product_name']
prod_subset = prod.loc[condition, columns]
prod_subset.head(20)

product_id            product_name
13175       13176  Bag of Organic Bananas
16796       16797            Strawberries
21136       21137    Organic Strawberries
21902       21903    Organic Baby Spinach
24851       24852                  Banana
26208       26209                   Limes
27844       27845      Organic Whole Milk
47208       47209    Organic Hass Avocado
47625       47626             Large Lemon
47765       47766         Organic Avocado

In [55]:
prod_subset = prod.loc[condition, columns]
prod_subset.head(20)

product_id            product_name
13175       13176  Bag of Organic Bananas
16796       16797            Strawberries
21136       21137    Organic Strawberries
21902       21903    Organic Baby Spinach
24851       24852                  Banana
26208       26209                   Limes
27844       27845      Organic Whole Milk
47208       47209    Organic Hass Avocado
47625       47626             Large Lemon
47765       47766         Organic Avocado

In [58]:
order_prod[
(order_prod['product_id'] == 13176) | 
(order_prod['product_id'] == 16797) | 
(order_prod['product_id'] == 21137) | 
(order_prod['product_id'] == 21903) | 
(order_prod['product_id'] == 24852) | 
(order_prod['product_id'] == 26209) | 
(order_prod['product_id'] == 27845) | 
(order_prod['product_id'] == 47209) | 
(order_prod['product_id'] == 47626) | 
(order_prod['product_id'] == 47766)].head(20)

order_id  product_id  add_to_cart_order  reordered
5           1       13176                  6          0
6           1       47209                  7          0
35         98       13176                  4          1
85        112       47209                  5          0
94        170       13176                  3          1
104       170       21137                 13          0
115       226       24852                  2          0
143       393       16797                  6          1
154       456       47626                  9          0
156       473       24852                  2          0
163       631       21137                  1          1
181       762       21137                  1          1
196       878       24852                  2          1
199       878       21903                  5          1
201       878       47209                  7          1
205       878       21137                 11          1
216       878       26209                 22          0
272      1042       24852                  1          1
280      1077       13176                  1          1
283      1077       21137                  4          1

In [86]:
condition = ((order_prod['product_id'] == 13176) | 
(order_prod['product_id'] == 16797) | 
(order_prod['product_id'] == 21137) | 
(order_prod['product_id'] == 21903) | 
(order_prod['product_id'] == 24852) | 
(order_prod['product_id'] == 26209) | 
(order_prod['product_id'] == 27845) | 
(order_prod['product_id'] == 47209) | 
(order_prod['product_id'] == 47626) | 
(order_prod['product_id'] == 47766))

columns = ['product_id']

order_subset = order_prod.loc[condition, columns]
order_subset.head(20)

product_id
5         13176
6         47209
35        13176
85        47209
94        13176
104       21137
115       24852
143       16797
154       47626
156       24852
163       21137
181       21137
196       24852
199       21903
201       47209
205       21137
216       26209
272       24852
280       13176
283       21137

In [88]:
#now, finally we can merge!
#used how='inner' first and did not work, changed it to outer and it showed a table
merged = pd.merge(prod_subset, order_subset, how='outer', on='product_id')

merged.head()

product_id            product_name
0       13176  Bag of Organic Bananas
1       13176  Bag of Organic Bananas
2       13176  Bag of Organic Bananas
3       13176  Bag of Organic Bananas
4       13176  Bag of Organic Bananas

In [90]:
merged.product_name.value_counts()
#Here is the answer!

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
import seaborn as sns
%matplotlib inline
import pandas as pd
import numpy as np
flights = sns.load_dataset('flights')

In [105]:
##### YOUR CODE HERE #####
flights.head(5)

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [0]:
columns = flights['month']


In [0]:
index = flights['year']


In [110]:
flights.columns.tolist()


['year', 'month', 'passengers']

In [111]:
flights.index.tolist()


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143]

In [118]:
flights = flights.reset_index()
flights

ValueError: ignored

In [119]:
tidy = flights.melt(id_vars='index', value_vars=['month', 'passengers'])
tidy

index    variable      value
0        0       month    January
1        1       month   February
2        2       month      March
3        3       month      April
4        4       month        May
5        5       month       June
6        6       month       July
7        7       month     August
8        8       month  September
9        9       month    October
10      10       month   November
11      11       month   December
12      12       month    January
13      13       month   February
14      14       month      March
15      15       month      April
16      16       month        May
17      17       month       June
18      18       month       July
19      19       month     August
20      20       month  September
21      21       month    October
22      22       month   November
23      23       month   December
24      24       month    January
25      25       month   February
26      26       month      March
27      27       month      April
28      28       month        May
29      29       month       June
..     ...         ...        ...
258    114  passengers        491
259    115  passengers        505
260    116  passengers        404
261    117  passengers        359
262    118  passengers        310
263    119  passengers        337
264    120  passengers        360
265    121  passengers        342
266    122  passengers        406
267    123  passengers        396
268    124  passengers        420
269    125  passengers        472
270    126  passengers        548
271    127  passengers        559
272    128  passengers        463
273    129  passengers        407
274    130  passengers        362
275    131  passengers        405
276    132  passengers        417
277    133  passengers        391
278    134  passengers        419
279    135  passengers        461
280    136  passengers        472
281    137  passengers        535
282    138  passengers        622
283    139  passengers        606
284    140  passengers        508
285    141  passengers        461
286    142  passengers        390
287    143  passengers        432

[288 rows x 3 columns]

In [120]:
tidy = tidy.rename(columns={
    'index': 'year', 
    'variable': 'month', 
    'value': 'passengers'
})
tidy

year       month passengers
0       0       month    January
1       1       month   February
2       2       month      March
3       3       month      April
4       4       month        May
5       5       month       June
6       6       month       July
7       7       month     August
8       8       month  September
9       9       month    October
10     10       month   November
11     11       month   December
12     12       month    January
13     13       month   February
14     14       month      March
15     15       month      April
16     16       month        May
17     17       month       June
18     18       month       July
19     19       month     August
20     20       month  September
21     21       month    October
22     22       month   November
23     23       month   December
24     24       month    January
25     25       month   February
26     26       month      March
27     27       month      April
28     28       month        May
29     29       month       June
..    ...         ...        ...
258   114  passengers        491
259   115  passengers        505
260   116  passengers        404
261   117  passengers        359
262   118  passengers        310
263   119  passengers        337
264   120  passengers        360
265   121  passengers        342
266   122  passengers        406
267   123  passengers        396
268   124  passengers        420
269   125  passengers        472
270   126  passengers        548
271   127  passengers        559
272   128  passengers        463
273   129  passengers        407
274   130  passengers        362
275   131  passengers        405
276   132  passengers        417
277   133  passengers        391
278   134  passengers        419
279   135  passengers        461
280   136  passengers        472
281   137  passengers        535
282   138  passengers        622
283   139  passengers        606
284   140  passengers        508
285   141  passengers        461
286   142  passengers        390
287   143  passengers        432

[288 rows x 3 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####